components/lock/in-memory/in_memory_lock.go

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2021 Layotto Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package in_memory
+
+import (
+	"go.uber.org/atomic"
+	"mosn.io/layotto/components/lock"
+	"sync"
+	"time"
+)
+
+type InMemoryLock struct {
+	features []lock.Feature
+	data     *sync.Map
+	wLock    sync.Mutex
+}
+
+type lockData struct {
+	key        string
+	owner      string
+	expireTime time.Time
+	lock       *atomic.Int32
+}
+
+func NewInMemoryLock() *InMemoryLock {
+	return &InMemoryLock{
+		features: make([]lock.Feature, 0),
+		data:     &sync.Map{},
+	}
+}
+
+func (s *InMemoryLock) Init(_ lock.Metadata) error {
+	return nil
+}
+
+func (s *InMemoryLock) Features() []lock.Feature {
+	return s.features
+}
+
+func (s *InMemoryLock) TryLock(req *lock.TryLockRequest) (*lock.TryLockResponse, error) {
+	item, ok := s.data.Load(req.ResourceId)
+	if !ok {
+		newItem := &lockData{
+			key:  req.ResourceId,
+			lock: &atomic.Int32{},
+		}
+		s.wLock.Lock()
```

Comment:
I remember that sync.Map is concurrency-safe. There seems no need to use a lock?

Reply:
> I remember that sync.Map is concurrency-safe. There seems no need to use a lock?

Line 73-78, has delete and store operation, so I locked it here.